fdb-extensions/src/test/java/com/apple/foundationdb/async/RangeSetTest.java

```diff
@@ -233,66 +240,182 @@ public void insert() {
+
+            // Should fail
+            final int failAfter = currTr;
+            assertFalse(rs.contains(multi.get(currTr++), new byte[]{0x03}).join());
+            assertFalse(rs.contains(multi.get(currTr++), new byte[]{0x03, 0x01}).join());
+            assertTrue(rs.contains(multi.get(currTr++), new byte[]{0x02}).join()); // fails because the key 0x02 itself is updated
+            assertEquals(2, rs.missingRanges(multi.get(currTr++), new Range(new byte[]{0x01}, new byte[]{0x05}))
+                    .asList().join().size());
+            assertEquals(1, rs.missingRanges(multi.get(currTr++), new Range(new byte[]{0x01}, new byte[]{0x02, 0x01}))
+                    .asList().join().size());
+            assertEquals(0, rs.missingRanges(multi.get(currTr++), r1)
+                    .asList().join().size());
+            assertEquals(1, rs.missingRanges(multi.get(currTr++), new Range(new byte[]{0x03, 0x01}, new byte[]{0x04, 0x01}))
+                    .asList().join().size());
+            assertFalse(rs.isEmpty(multi.get(currTr++)).join());
+
+            assertEquals(multi.size(), currTr);
+            multi.commit(IntStream.range(0, multi.size()).filter(i -> i >= failAfter).boxed().collect(Collectors.toSet()));
+        }
+    }
+
+    @Test
+    void failIfKeyInRangeIsInsertedConcurrently() {
+        try (MultipleTransactions multi = MultipleTransactions.create(db, 2)) {
+            Range r = new Range(new byte[]{0x01}, new byte[]{0x02});
+            Range overlappingRange = new Range(new byte[]{0x00, 0x01}, new byte[]{0x01, 0x01});
+
+            // Insert a range into the set
+            Transaction tr1 = multi.get(0);
+            assertTrue(rs.insertRange(tr1, overlappingRange, true).join());
+
+            // Read from a range that overlaps this range
+            Transaction tr2 = multi.get(1);
+            assertEquals(List.of(r), rs.missingRanges(tr2, r).asList().join());
+
+            // The second transaction should fail to commit
+            multi.commit(Set.of(1));
+        }
+    }
+
+    @Test
+    void insertNonOverlappingRanges() {
+        try (MultipleTransactions multi = MultipleTransactions.create(db, 10)) {
+            for (int i = 0; i < multi.size(); i++) {
+                Transaction tr = multi.get(i);
+                assertTrue(rs.insertRange(tr, new Range(new byte[]{(byte)i}, new byte[]{(byte)i, 0x00}), i % 2 == 0).join());
+            }
+
+            // All transactions should be able to commit
+            multi.commit();
+        }
+    }
+
+    @Test
+    void insertOverlappingRanges() {
+        try (MultipleTransactions multi = MultipleTransactions.create(db, 10)) {
+            for (int i = 0; i < multi.size(); i++) {
+                Transaction tr = multi.get(i);
+                assertTrue(rs.insertRange(tr, new Range(new byte[]{(byte)i}, new byte[]{(byte)(i + 1), 0x00}), false).join());
+            }
+
+            // Even transactions should be able to commit
```

Comment:
even transactions can commit because it commits in order, so in chronological order we have:
1. tr0 commits
2. tr1 conflicts with tr0
3. tr2 does not overlap with tr0's range, so commits
4. tr3 conflicts with tr2
5. tr4 does not overlap with the two that committed, so succeeds
...

Reply:
Yeah, that's right. Expanded this comment